# Release 3.0.0

golluroSICKAG opened the pull request:
# Release 3.0.0

## New features
- Provide features even without CSK_IODDInterpreter
- Possibility to create readMessages without IODD information, by setting start-/endByte and unpack format
- Possibility to search and cut relevant part out of created JSON within readMessage
- New FlowConfig block 'OnNewDataAuto' which automatically creates readMessages and optionally checks to power selected port (WARNING: Do NOT mix with manual readMessage setup)
- React on "OnStopFlowConfigProviders" event of FlowConfig modul to stop pulling IO-Link data
- Check and handle different kind of IOLink.SMI APIs
- Function to delete all existing readMessages
- Check if persistent data to load provides all relevant parameters. Otherwise add default values
- Selectable if timers for readMessages should start automatically after parameters were loaded


## Improvements
- Changed event name "OnNewReadMessage_PORT_MESSAGENAME" to "OnNewReadMessage_INSTANCE_PORT_MESSAGENAME" to make it possible to switch between ports during runtime
- Do not start readMessage timers automatically with creation of readMessage
- Use new event 'OnNewRawReadMessage_INSTANCE_PORT_MESSAGENAME" within FlowConfig (only providing data content as first parameter)
- If using FlowConfig, start readMessage timers after 5 seconds
- Add info about port within "OnNewIOLinkPortStatus" event
- Add list of ports related to instances within "getInstancePortMap" function
- Better handling if CSK_IODDInterpreter is not available

## Bugfix
- Issue with CSK_UserManagement support
- Error if sending process data without any selection
- Error in handling IODD data

## Following tasks were checked

- [x] Module is coded and structured according to the "Developing guideline for modules" described within the CSK documentation
- [x] All functions/events/parameters are documented within the manifest documentation
- [x] The manifest description of the main CROWN includes main information about the purpose of the module and how to use it in general
- [x] API docu based on the manifest was created and stored within the "docu"-folder of the repository
- [x] Internal LUA code documentation exists for variables and non served functions
- [x] All relevant infos are logged via the SharedLogger 'ModuleLogger'
- [x] Module supports persistent data feature based on 'CSK_Module_PersistentData'
- [x] Module supports user management based on 'CSK_Module_UserManagement'
- [x] No open "ToDos" within the code or at least clearly explained comments why they exist...
- [x] "Version" key in app manifest was updated following semantic versioning (and use '0.x.y' for test / experimental modules which are not yet ready to be officially released)
- [x] Meaningful IDs used for UI elements
- [x] Module was tested on an AppSpace device (at least on SICK AppEngine) with no error message
- [x] README.md is up to date (incl. info of device + firmware the module was tested with)
- [x] CHANGELOG.md is up to date

golluroSICKAG committed:
# Release 3.0.0

## New features
- Provide features even without CSK_IODDInterpreter
- Possibility to create readMessages without IODD information, by setting start-/endByte and unpack format
- Possibility to search and cut relevant part out of created JSON within readMessage
- New FlowConfig block 'OnNewDataAuto' which automatically creates readMessages and optionally checks to power selected port (WARNING: Do NOT mix with manual readMessage setup)
- React on "OnStopFlowConfigProviders" event of FlowConfig modul to stop pulling IO-Link data
- Check and handle different kind of IOLink.SMI APIs
- Function to delete all existing readMessages

## Improvements
- Changed event name "OnNewReadMessage_PORT_MESSAGENAME" to "OnNewReadMessage_INSTANCE_PORT_MESSAGENAME" to make it possible to switch between ports during runtime
- Do not start readMessage timers automatically with creation of readMessage
- Use new event 'OnNewRawReadMessage_INSTANCE_PORT_MESSAGENAME" within FlowConfig (only providing data content as first parameter)
- If using FlowConfig, start readMessage timers after 5 seconds
- Add info about port within "OnNewIOLinkPortStatus" event
- Add list of ports related to instances within "getInstancePortMap" function

## Bugfix
- Issue with CSK_UserManagement support

golluroSICKAG committed:
Merged updates of version 2.1.1 into this version

golluroSICKAG committed:
Merge branch 'main' into main

golluroSICKAG committed:
Fix wrong version

golluroSICKAG committed:
Edit in Changelog

golluroSICKAG committed:
- New feature to select specific bit for readMessage without IODD interpretation
- New feature to register to event to forward content via writeMessage
- New FlowConfig block to writeProcessData
- Error handling if CSK_IODDInterpreter is not available

golluroSICKAG committed:
* minor fix

golluroSICKAG committed:
Minor updates

teplymaSICKAG commented on CSK_Module_MultiIOLinkSMI/scripts/CSK_MultiIOLinkSMI_Processing.lua:
```diff
@@ -584,25 +706,33 @@ local function writeIODDMessage(messageName, jsonDataToWrite)
-      elseif dataMode == 'Parameters' then
-        success, errorCode = writeParameter(ioddWriteMessages[messageName].dataInfo.Parameters[dataPointID], dataPointDataToWrite)
+  if ioddWriteMessages[messageName] then
+    if ioddWriteMessages[messageName].dataInfo.ProcessData then
```
```suggestion
    if ioddWriteMessages[messageName].dataInfo then
```
Otherwise it is not possible to write only ISDU (parameter) without process data

golluroSICKAG replied:
merged

teplymaSICKAG commented on CSK_Module_MultiIOLinkSMI/scripts/CSK_MultiIOLinkSMI_Processing.lua:
```diff
@@ -504,21 +596,29 @@ local function updateIODDReadMessages()
+      else
+        success, messageContent = readIODDMessage(messageName)
+        ioddReadMessagesResults[messageName] = success
+        ioddLatestReadMessages[messageName] = jsonMessageContent
```
```suggestion
        ioddLatestReadMessages[messageName] = messageContent
```
Otherwise the refreshing live payload button doesn't work

golluroSICKAG replied:
merged

teplymaSICKAG reviewed (changes requested):
These changes are small but important :)

golluroSICKAG committed:
Fixes after review

golluroSICKAG committed:
- Check if persistent data to load provides all relevant parameters. Otherwise add default values
- Selectable if timers for readMessages should start automatically after parameters were loaded

golluroSICKAG commented:
thx for the review!
